File: src/main/java/org/code4everything/hutool/converter/ArrayConverter.java
package org.code4everything.hutool.converter;

import cn.hutool.core.util.ArrayUtil;
import org.code4everything.hutool.CliException;
import org.code4everything.hutool.Converter;
import org.code4everything.hutool.Hutool;
import org.code4everything.hutool.MethodArg;
import org.code4everything.hutool.Utils;

import java.lang.reflect.Array;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Objects;

/**
 * @author pantao
 * @since 2021/6/24
 */
public class ArrayConverter implements Converter<Object> {

    private Class<?> componentClass;

    public ArrayConverter(Class<?> arrayClass) {
        setComponentClass(arrayClass);
    }

    public void setComponentClass(Class<?> arrayClass) {
        Objects.requireNonNull(arrayClass);
        componentClass = arrayClass.getComponentType();
        if (componentClass == null || componentClass.isArray()) {
            String msg = "convert array error: only support linear array, cannot convert multidimensional array";
            Hutool.debugOutput(msg);
            throw new CliException(msg);
        }
    }

    @Override
    public Object string2Object(String string) {
        if (Utils.isStringEmpty(string)) {
            return Array.newInstance(componentClass, 0);
        }
        List<String> segments = Arrays.asList(string.split(MethodArg.getSeparator()));
        Object array = Array.newInstance(componentClass, segments.size());
        for (int i = 0; i < segments.size(); i++) {
            Array.set(array, i, Hutool.castParam2JavaType(null, segments.get(i), componentClass, false));
        }
        return array;
    }

    @Override
    public String object2String(Object object) throws Exception {
        if (!ArrayUtil.isArray(object)) {
            return "";
        }
        int length = Array.getLength(object);
        List<Object> list = new ArrayList<>();
        for (int i = 0; i < length; i++) {
            list.add(Array.get(object, i));
        }
        return ArrayUtil.isArray(object) ? new ListStringConverter().object2String(list) : "";
    }
}

File: src/main/kotlin/org/code4everything/hutool/converter/ArrayConverter.kt
package org.code4everything.hutool.converter

import java.lang.reflect.Array
import org.code4everything.hutool.CliException
import org.code4everything.hutool.Converter
import org.code4everything.hutool.Hutool
import org.code4everything.hutool.MethodArg
import org.code4everything.hutool.Utils

class ArrayConverter(arrayClass: Class<*>) : Converter<Any> {

    private val msg = "convert array error: only support linear array, cannot convert multidimensional array"

    private lateinit var componentClass: Class<*>

    fun setComponentClass(arrayClass: Class<*>) {
        componentClass = arrayClass.componentType.also { if (it == null || it.isArray) throw CliException(msg) }
    }

    override fun string2Object(string: String): Any {
        if (Utils.isStringEmpty(string)) {
            return Array.newInstance(componentClass, 0)
        }
        val segments = string.split(MethodArg.separator)
        return Array(segments.size) { i -> Hutool.castParam2JavaType(null, segments[i], componentClass, false) }
    }

    override fun object2String(any: Any?): String {
        return if (any is kotlin.Array<*> || any is Array) convert(any) else ""
    }

    private fun convert(any: Any): String {
        val list = List(Array.getLength(any)) { i -> Array.get(any, i) }
        return ListStringConverter().object2String(list)
    }

    init {
        setComponentClass(arrayClass)
    }
}
